File: 3sum/3sum.cpp
class Solution {
public:
    vector<vector<int>> threeSum(vector<int>& nums) {
        int j;
        int k;
        int sum;
        vector<vector<int> > ans;
        sort(nums.begin(), nums.end());
        if(nums.size() < 3) return ans;
        for(int i=0; i<nums.size()-2;i++){
            if(i==0 || (i>0 && nums[i] != nums[i-1])) {
                j = i+1;
                k = nums.size()-1;
                
                while(j<k){
                    sum = nums[i] + nums[j] + nums[k];
                    if(sum > 0){
                        k--;
                    }
                    else if(sum < 0){
                        j++;
                    }
                    else if(sum == 0){
                        ans.push_back({nums[i],nums[j],nums[k]});
                        while(j<k && nums[j] == nums[j+1]) j++;
                        while(j<k && nums[k] == nums[k-1]) k--;

                        j++;
                        k--;
                    }    
                }
            }
        }
        return ans;
    }
};

File: 3sum/3sum.py
class Solution(object):
    def threeSum(self, nums):
        """
        :type nums: List[int]
        :rtype: List[List[int]]
        """
        nums.sort()
        ans = set()
        for i in range(len(nums)):
            if i==0 or (i>0 and nums[i]!=nums[i-1]):
                j = i+1
                k = len(nums)-1
                while(j<k):
                    tmp = nums[i] + nums[j] + nums[k]
                    if tmp==0:
                        ans.add((nums[i],nums[j],nums[k]))
                        while(j<k and nums[j]==nums[j+1]): j+=1
                        while(j<k and nums[k]==nums[k-1]): k-=1
                        j+=1
                        k-=1
                    elif tmp>0:
                        k -= 1
                    else:
                        j += 1
        return ans
